### DISCUSSION TITLE
ScrollBar AllowAutoHide delay

### DISCUSSION THREAD

**Discussion by sinatrocious** [AvaloniaUI/Avalonia#14311] [Q&A] [answered]
Is there a way to adjust `ScrollBar` auto-hide/auto-show delays?

As of now user have to position mouse cursor *over* scrollbar to auto-show it and wait a bit for it to being shown. Then he can use it. But there are no hints as to where to stop the mouse: stopping too early or moving too far and scroll bar will not be shown.

The easist workaround will be to show scroll bar *instantly*, then it's much easier to target it and not going too far. I'd also prefer to hide it asap, once the mouse is not over it.

---

I have problem with scroll bar, located too near to the right edge of the resizable window. When the user is trying to mouse-over scroll bar, he tries to move it precisely *over* that very thin scroll bar and often miss (move too much to the right), have its cursor changed to horizontal resize and it's pretty terrible user experience. There is no way to know he can stop earlier, wait and scrollbar will be shown.

This can be reproduced easily by making new avalonia desktop application and making greetings `TextBlock` big and putting it into a `ScrollViewer` in the `MainView.xaml`:

```xaml
<ScrollViewer>
    <TextBlock Text="{Binding Greeting}"
               HorizontalAlignment="Center"
               VerticalAlignment="Center"
               Height="1000"/>
</ScrollViewer>
```

![image](https://github.com/AvaloniaUI/Avalonia/assets/91904521/8900cbb9-af9c-45c9-bafa-e1d0c5d90cb2)

### Some thoughts about delays:

In my opinion those delay are bad idea. They are useful with e.g. tooltips to prevent accidental showing of tooltips when moving mouse over multiple elements with tooltips. If tooltips are *big* and positioned at mouse cursor that simply would not looks nice: poping up tooltips overlapping otherlying elements. But scroll bar is different: if mouse is accidentally moved over and scrool bar is auto-shown there is not problem, since scrollbars are small and don't overlap anything.

**Comment by maxkatz6** [comment]
It seems to be hardcoded in couple of transitions of ScrollBar style:
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/Controls/ScrollBar.xaml#L206

It's probably easier to copy paste this ControlTheme, with changed transitions duration, instead of trying to re-style each of them independently.

**Reply by sinatrocious** [reply]
Close, but that's not all what is happening on mouseover, `AllowAutoHide` is processed elsewhere.

Here I've changed `Duration` to `"0:0:0.6"`, so you see it's not synced, but there is a delay before that.

![avalonia scrollbar delays](https://github.com/AvaloniaUI/Avalonia/assets/91904521/612031a0-8e02-47d9-b39a-ea3b916e5732)

**Comment by msneijders** [comment, marked as answer]
I think you're looking for `ScrollBar.HideDelay` (default at 2 secs) and `ScrollBar.ShowDelay`? (default at 0.5 secs). A custom controltheme with your own values for these is still most practical probably.

**Reply by sinatrocious** [reply]
Yes, that's it! Thanks. After setting in style both delays to `0` the `ScrollBar` is now very easy to use.

Hmm.. setting delays to `0` worked, but to anything else will bug `Scrollbar`. Good for me I only want `0`.

**Reply by sinatrocious** [reply]
Those properties not yet [documented](https://reference.avaloniaui.net/api/Avalonia.Controls.Primitives/ScrollBar). And I was wondering, why I didn't find those properties while searching for "delay".

Also, setting `ScrollBar.Width` leads to some very ugly result:

<details>
  <summary>Show ugly result (scaled)</summary>

<img src="https://github.com/AvaloniaUI/Avalonia/assets/91904521/12a4c6e2-9d03-420b-87aa-d90bea9c26ac" width="200">

</details>
